### PR TITLE
[WIP] traits/select: use global vs per-infcx caches more uniformly.

### PR DESCRIPTION
**Pull request by eddyb**
~~*Note: this is based on an older `master` to avoid perf interference before https://github.com/rust-lang/rust/pull/67953#issuecomment-587031702 is resolved.*~~ **EDIT**: sadly not workable due https://github.com/rust-lang/rust/pull/69294#issuecomment-588360527.

So far this PR only has the first couple steps towards making the decision of which cache ("global" vs "per-`InferCtxt`") to use for trait evaluation/selection, only once (at the time of checking the cache).

The goal here is to actually make per-`InferCtxt` caches not track `DepNode`s, and maybe even enforce that once `SelectionContext::in_task` is entered, the `InferCtxt` is effectively unused.

My assumption is that if you *need* inference variables in your cache key (i.e. `ParamEnv` and/or `PolyTraitPredicate`) to ever end up doing anything "non-global", and you can't get there from a "global cache key" (which would still use `DepNode`s and `in_task` etc.).

Perhaps another path forward is to redirect "global cache keys" to a query, but I'm not sure how that would handle cycles (badly?), and it feels like stepping on Chalk's toes.

r? @nikomatsakis cc @rust-lang/wg-traits @Zoxc @michaelwoerister

### CODE REVIEW

**Comment by eddyb** [comment]
@Mark-Simulacrum Ugh, I was hoping to do this on top of 58b834344fc7b9185e7a50db1ff24e5eb07dae5e, before all of the recent `rustc::{infer,traits}` changes, but `bors` seems to want to merge against current master.

**Comment by Mark-Simulacrum** [comment]
> Ugh, I was hoping to do this on top of [58b8343](https://github.com/rust-lang/rust/commit/58b834344fc7b9185e7a50db1ff24e5eb07dae5e), before all of the recent `rustc::{infer,traits}` changes, but `bors` seems to want to merge against current master.

AFAIK, if you really want this you can add revert commits to this branch for master's commits since then and then compare with an older base commit (vs. using the link perf auto-generates).

**Comment by eddyb** [comment]
@Mark-Simulacrum I was shocked at how far down I had to go in `git log` to find the PRs I wanted to base myself just before, I don't think I could seriously consider reverting all of them, heh.

**Comment by Mark-Simulacrum** [comment]
In theory `git revert $(git log <base PR bors commit> --author=bors --format %H)` should work, but I haven't tested :)

**Comment by eddyb** [comment]
It's more that I would find that excessive and it would make a mess of the PR.

**Comment by eddyb** [comment]
Surprised this is already a win, this is good news (results are available, idk why the bot didn't post another comment).

File: src/librustc_infer/traits/select.rs [hunk 1]
```diff
@@ -1285,9 +1277,7 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
         result: &SelectionResult<'tcx, SelectionCandidate<'tcx>>,
     ) -> bool {
         match result {
-            Ok(Some(SelectionCandidate::ParamCandidate(trait_ref))) => {
-                !trait_ref.skip_binder().input_types().any(|t| t.walk().any(|t_| t_.is_ty_infer()))
-            }
+            Ok(Some(SelectionCandidate::ParamCandidate(trait_ref))) => !trait_ref.needs_infer(),
```
Comment (eddyb) [inline src/librustc_infer/traits/select.rs]: This is a fix for https://github.com/rust-lang/rust/pull/55258#discussion_r383309878, I should make a separate commit but I didn't want to lose track of this.

**Comment by nikomatsakis** [comment]
BTW I'm planning to review this today or tomorrow

File: src/librustc_infer/traits/select.rs [hunk 2]
```diff
@@ -1250,18 +1240,13 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
...
-            .borrow()
-            .get(&param_env.and(*trait_ref))
-            .map(|v| v.get(tcx))
+        let cache = if self.can_use_global_caches(param_env) && !trait_ref.has_local_value() {
+            &tcx.selection_cache
+        } else {
+            &self.infcx.selection_cache
+        };
+
+        cache.hashmap.borrow().get(&param_env.and(*trait_ref)).map(|v| v.get(tcx))
```
Comment (nikomatsakis) [inline src/librustc_infer/traits/select.rs]: to be clear, this commit is just a "logical cleanup", right?

File: src/librustc_infer/traits/select.rs [hunk 3]
```diff
@@ -1309,31 +1295,35 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
...
-                    // This may overwrite the cache with the same value.
-                    tcx.selection_cache
-                        .hashmap
-                        .borrow_mut()
-                        .insert(param_env.and(trait_ref), WithDepNode::new(dep_node, candidate));
-                    return;
-                }
-            }
+        // HACK(eddyb) never cache overflow (this check used to be global-only).
+        if let Err(Overflow) = candidate {
```
Comment (nikomatsakis) [inline src/librustc_infer/traits/select.rs]: with the exception of this tiny logic change, I guess

Comment (nikomatsakis) [reply src/librustc_infer/traits/select.rs]: I can't really think why we would want to cache overflow *locally* anyhow, that seems like a bug...

**Review by nikomatsakis — APPROVED** [review]
This all makes a lot of sense to me thus far. Sorry for being a bit slow.

File: src/librustc_infer/traits/select.rs (hunk 3 above)
Comment (eddyb) [reply src/librustc_infer/traits/select.rs]: Pfft, that's nothing, here's something more fun: I don't think we ever use the local caches except for the `intercrate` mode, because we're doing freshening.

So we could put `intercrate` in `Reveal` or w/e then replace freshening with canonicalization, and have only global caching (either that or I'm seriously misunderstanding the implementation).

**Comment by nikomatsakis** [comment]
@eddyb is this PR considered "eligible to merge" or is it blocked on something?

**Comment by eddyb** [comment]
@nikomatsakis This is waiting for M1 from https://github.com/w3f/General-Grants-Program/pull/261 (i.e. self-profiling integration for the trait system), so that we can measure the impact of this. I'll likely open a PR for that soon.

**Comment by Dylan-DPC** [comment]
@eddyb the grants PR has been merged, is there anything else this blocked on?

**Comment by nikomatsakis** [comment]
I'm going to close this PR for inactivity. @eddyb re-open if you want when you think it's ready to go forward
